MethodHandleNatives.resolve should not verify access if caller is null

**Describe the issue**
`MethodHandles.Lookup.unreflectSetter` fails with:

```
Caused by: java.lang.NoSuchMethodError: jdk.internal.misc.Unsafe.putReference(java.lang.Object, long, java.lang.Object)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.methodhandles.Util_java_lang_invoke_MethodHandleNatives.resolve(Target_java_lang_invoke_MethodHandleNatives.java:347)
	at java.base@17.0.4-internal/java.lang.invoke.MethodHandleNatives.resolve(MethodHandleNatives.java:223)
	at java.base@17.0.4-internal/java.lang.invoke.MemberName$Factory.resolve(MemberName.java:1085)
	at java.base@17.0.4-internal/java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1114)
	... 10 more
```

Method handles logic within GraalVM should be setting `Unsafe.putReference` for reflection access. Where would be the best place to add that?

However, even if you manually add the method for reflection access, you get:

```
Exception in thread "main" org.graalvm.compiler.debug.GraalError: java.lang.reflect.InvocationTargetException
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.methodhandles.Util_java_lang_invoke_MethodHandleNatives.verifyAccess(Target_java_lang_invoke_MethodHandleNatives.java:370)
	at java.base@17.0.4-internal/java.lang.invoke.MethodHandleNatives.resolve(MethodHandleNatives.java:226)
	at java.base@17.0.4-internal/java.lang.invoke.MemberName$Factory.resolve(MemberName.java:1085)
	at java.base@17.0.4-internal/java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1114)
	at java.base@17.0.4-internal/java.lang.invoke.DirectMethodHandle.makePreparedFieldLambdaForm(DirectMethodHandle.java:775)
	at java.base@17.0.4-internal/java.lang.invoke.DirectMethodHandle.preparedFieldLambdaForm(DirectMethodHandle.java:692)
	at java.base@17.0.4-internal/java.lang.invoke.DirectMethodHandle.preparedFieldLambdaForm(DirectMethodHandle.java:680)
	at java.base@17.0.4-internal/java.lang.invoke.DirectMethodHandle.make(DirectMethodHandle.java:113)
	at java.base@17.0.4-internal/java.lang.invoke.DirectMethodHandle.make(DirectMethodHandle.java:129)
	at java.base@17.0.4-internal/java.lang.invoke.MethodHandles$Lookup.getDirectFieldCommon(MethodHandles.java:4049)
	at java.base@17.0.4-internal/java.lang.invoke.MethodHandles$Lookup.getDirectFieldNoSecurityManager(MethodHandles.java:4040)
	at java.base@17.0.4-internal/java.lang.invoke.MethodHandles$Lookup.unreflectField(MethodHandles.java:3504)
	at java.base@17.0.4-internal/java.lang.invoke.MethodHandles$Lookup.unreflectSetter(MethodHandles.java:3487)
	at org.example.methodhandles.MethodHandling.main(MethodHandling.java:17)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base@17.0.4-internal/java.lang.reflect.Method.invoke(Method.java:568)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.methodhandles.Util_java_lang_invoke_MethodHandleNatives.verifyAccess(Target_java_lang_invoke_MethodHandleNatives.java:368)
	... 13 more
Caused by: java.lang.NullPointerException
	at java.base@17.0.4-internal/sun.invoke.util.VerifyAccess.isSamePackage(VerifyAccess.java:376)
	at java.base@17.0.4-internal/sun.invoke.util.VerifyAccess.isClassAccessible(VerifyAccess.java:195)
	at java.base@17.0.4-internal/sun.invoke.util.VerifyAccess.isMemberAc,essible(VerifyAccess.java:104)
	... 15 more
```

I don't think `Util_java_lang_invoke_MethodHandleNatives.resolve` is implementing things correctly. Looking at the JDK implementation of that method, it shows that verifying class access is only done when `caller` class is not null.

```
// void resolve(MemberName self, Class<?> caller)
JVM_ENTRY(jobject, MHN_resolve_Mem(JNIEnv *env, jobject igcls, jobject mname_jh, jclass caller_jh,
    jint lookup_mode, jboolean speculative_resolve)) {
  if (mname_jh == NULL) { THROW_MSG_NULL(vmSymbols::java_lang_InternalError(), "mname is null"); }
  Handle mname(THREAD, JNIHandles::resolve_non_null(mname_jh));

  // The trusted Java code that calls this method should already have performed
  // access checks on behalf of the given caller.  But, we can verify this.
  // This only verifies from the context of the lookup class.  It does not
  // verify the lookup context for a Lookup object teleported from one module
  // to another. Such Lookup object can only access the intersection of the set
  // of accessible classes from both lookup class and previous lookup class.
  if (VerifyMethodHandles && (lookup_mode & LM_TRUSTED) == LM_TRUSTED && caller_jh != NULL &&
      java_lang_invoke_MemberName::clazz(mname()) != NULL) {
    Klass* reference_klass = java_lang_Class::as_Klass(java_lang_invoke_MemberName::clazz(mname()));
    if (reference_klass != NULL && reference_klass->is_objArray_klass()) {
      reference_klass = ObjArrayKlass::cast(reference_klass)->bottom_klass();
    }

    // Reflection::verify_class_access can only handle instance classes.
    if (reference_klass != NULL && reference_klass->is_instance_klass()) {
      // Emulate LinkResolver::check_klass_accessability.
      Klass* caller = java_lang_Class::as_Klass(JNIHandles::resolve_non_null(caller_jh));
      // access check on behalf of the caller if this is not a public lookup
      // i.e. lookup mode is not UNCONDITIONAL
      if ((lookup_mode & LM_UNCONDITIONAL) == 0
          && Reflection::verify_class_access(caller,
                                             InstanceKlass::cast(reference_klass),
                                             true) != Reflection::ACCESS_OK) {
        ResourceMark rm(THREAD);
        stringStream ss;
        ss.print("caller %s tried to access %s", caller->class_in_module_of_loader(),
                 reference_klass->class_in_module_of_loader());
        THROW_MSG_NULL(vmSymbols::java_lang_InternalError(), ss.as_string());
      }
    }
  }
...
```

Adding `caller != null` condition to the if statement before `verifyAccess` call appears to be enough to fix the issue. I'll be sending a PR shortly.

**Steps to reproduce the issue**
Please include both build steps as well as run steps
1. **git clone --depth 1 https://github.com/galderz/mendrugo**
2. **cd methodhandles**
3. **make**

**Describe GraalVM and your environment:**
 - GraalVM version 2921da019434
 - JDK major version: **17**
 - OS: **macOS**
 - Architecture: **ARM64**

--- a/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/methodhandles/Target_java_lang_invoke_MethodHandleNatives.java
+++ b/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/methodhandles/Target_java_lang_invoke_MethodHandleNatives.java
@@ -222,7 +222,7 @@ final class Target_java_lang_invoke_MethodHandleNatives {
         Class<?> declaringClass = self.getDeclaringClass();
         Target_java_lang_invoke_MemberName resolved = Util_java_lang_invoke_MethodHandleNatives.resolve(self, caller, speculativeResolve);
         assert resolved == null || resolved.reflectAccess != null || resolved.intrinsic != null;
-        if (resolved != null && resolved.reflectAccess != null &&
+        if (resolved != null && resolved.reflectAccess != null && caller != null &&
                         !Util_java_lang_invoke_MethodHandleNatives.verifyAccess(declaringClass, resolved.reflectAccess.getDeclaringClass(), resolved.reflectAccess.getModifiers(), caller,
                                         lookupMode)) {
             throw new IllegalAccessError(resolved + " is not accessible from " + caller);

--- a/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/methodhandles/MethodHandleFeature.java
+++ b/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/methodhandles/MethodHandleFeature.java
@@ -26,6 +26,7 @@ package com.oracle.svm.hosted.methodhandles;
 
 import java.lang.invoke.CallSite;
 import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
 import java.lang.invoke.MethodType;
 import java.lang.reflect.Array;
 import java.lang.reflect.Field;
@@ -35,6 +36,7 @@ import java.util.Iterator;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 
+import jdk.internal.misc.Unsafe;
 import org.graalvm.compiler.serviceprovider.JavaVersionUtil;
 import org.graalvm.nativeimage.hosted.Feature;
 import org.graalvm.nativeimage.hosted.RuntimeReflection;
@@ -186,6 +188,9 @@ public class MethodHandleFeature implements Feature {
                         access.findClassByName("java.lang.invoke.VarHandle"));
 
         access.registerSubtypeReachabilityHandler(MethodHandleFeature::scanBoundMethodHandle, boundMethodHandleClass);
+
+        access.registerReachabilityHandler(MethodHandleFeature::registerUnsafePutReference,
+                        ReflectionUtil.lookupMethod(MethodHandles.Lookup.class, "unreflectSetter", Field.class));
     }
 
     private static void registerMHImplFunctionsForReflection(DuringAnalysisAccess access) {
@@ -260,6 +265,10 @@ public class MethodHandleFeature implements Feature {
                 }
             }
         }
+    }
+
+    private static void registerUnsafePutReference(DuringAnalysisAccess access) {
+        RuntimeReflection.register(ReflectionUtil.lookupMethod(Unsafe.class, "putReference", Object.class, long.class, Object.class));
     }
 
     private static String valueConverterName(Wrapper src, Wrapper dest) {

